### PR TITLE
[FLINK-5663] Prevent leaking SafetyNetCloseableRegistry though InheritableThreadLocal

### PR DESCRIPTION
**Pull request by StefanRRichter**
This PR prevents the `SafetyNetCloseableRegistry` from leaking into pooled threads through `InheritableThreadLocal`. 

As first step, we use `ThreadLocal` instead of `InheritableThreadLocal` to hold the closeable registries.

Additionally, we also create safety nets for the file system at the scope of the checkpointing thread. We hope that this covers already covers most cases. Other threads could actually also create safety nets for their scope right now.

As a last change, we made the reaper thread a singleton, because we could potentially create more registries now and it is not required to have one reaper thread per registry.

### CODE REVIEW

**Comment by StefanRRichter** [comment]
cc @StephanEwen

**Comment by StephanEwen** [comment]
Making the registry not inherited is a good quick fix.

How about moving the initialization and closing of that registry into the task's trigger checkpoint action? https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java#L1119

That way it is tied to the dedicated asynchronous thread, which is safer, I think. Having the registry initialization/closing in the `triggerCheckpoint(...)` method means the method can never be called by the main thread. It probably does not happen currently, but it seems quite easy to accidentally violate.

There is also some additional refactoring in this pull request that makes the reaper thread a static variable, shared across all registries. Currently that thread seems to never stopped by anyone. How about factoring out the changes for the single static reaper thread into a separate issue (that we do not merge for 1.2) and addressing the thread stopping as well (via an atomic count how many registries are currently open, stopping it when ging to zero, starting when going from zero).

**Comment by StephanEwen** [comment]
To clarify: The "static reaper thread" change is a good idea, but the change seems a bit incomplete without well defined shutdown of the thread.

**Comment by StefanRRichter** [comment]
I agree about the first part. 

About the reaper thread, I was also concerned about stopping and the counter is not a fix: there can be 0 registries but at some point a new task could come up, right?

The reaper is now a daemon thread, so it will go down with the JVM, before which all streams get closed anyways. Or am I missing something?

**Comment by StefanRRichter** [comment]
Ok, maybe what we are talking about is respawning the reaper thread if we go again from 0 to 1. That is possible.

**Comment by StephanEwen** [comment]
Concerning the reaper thread:
  - It is not really broken in the release branch, only has a few more threads then necessary.
  - So far we have strived to make sure Flink does not leave any lingering Threads at all (as validated by the MiniCluster thread), because it actually messes up testing setups from many users that repeatedly execute programs with the LocalEnvironment. That would be good to keep.
  - One can probably stop the thread when all registries are closed and re-spawn it when new registries come. That would be 

In summary: I think lingering threads are a type of regression, actually. Introducing that for something that is not broken right now is something I would not do for a release. Especially given that there is probably a cleaner solution that can both implement the improvement and not have the lingering threads regression. Let's do that for the 1.2.1/1.3 release.

**Comment by StefanRRichter** [comment]
Closed in favour of #3229 and #3230 .
